include/dd/Package.hpp

```diff
@@ -676,6 +679,201 @@ namespace dd {
+                if (root_edge.w.approximatelyZero()) {
+                    throw std::runtime_error("Numerical instabilities led to a 0-vector! Abort simulation!");
+                }
+                std::cerr << "WARNING in MAll: numerical instability occurred during simulation: |alpha|^2 + |beta|^2 = "
+                          << ComplexNumbers::mag2(root_edge.w) << ", but should be 1!\n";
+            }
+
+            vEdge            cur      = root_edge;
+            const QubitCount n_qubits = root_edge.p->v + 1;
+
+            std::string result(n_qubits, '0');
+
+            std::uniform_real_distribution<fp> dist(0.0, 1.0L);
+
+            for (Qubit i = root_edge.p->v; i >= 0; --i) {
+                fp p0  = ComplexNumbers::mag2(cur.p->e.at(0).w);
+                fp p1  = ComplexNumbers::mag2(cur.p->e.at(1).w);
+                fp tmp = p0 + p1;
+
+                if (std::abs(tmp - 1.0L) > epsilon) {
+                    throw std::runtime_error("Added probabilities differ from 1 by " + std::to_string(std::abs(tmp - 1.0L)));
+                }
+                p0 /= tmp;
+
+                const fp n = dist(mt);
+                if (n < p0) {
+                    cur = cur.p->e.at(0);
+                } else {
+                    result[cur.p->v] = '1';
+                    cur              = cur.p->e.at(1);
+                }
+            }
+
+            if (collapse) {
+                decRef(root_edge);
+
+                vEdge                e = vEdge::one;
+                std::array<vEdge, 2> edges{};
+
+                for (Qubit p = 0; p < n_qubits; p++) {
+                    if (result[p] == '0') {
+                        edges[0] = e;
+                        edges[1] = vEdge::zero;
+                    } else {
+                        edges[0] = vEdge::zero;
+                        edges[1] = e;
+                    }
+                    e = makeDDNode(p, edges, false);
+                }
+                incRef(e);
+                root_edge = e;
+                garbageCollect();
+            }
+
+            return std::string{result.rbegin(), result.rend()};
+        }
+
+    private:
+        double assignProbabilities(vEdge edge, std::unordered_map<vNode*, double>& probs) {
+            auto it = probs.find(edge.p);
+            if (it != probs.end()) {
+                return ComplexNumbers::mag2(edge.w) * it->second;
+            }
+            double sum;
+            if (edge.isTerminal()) {
+                sum = 1.0;
+            } else {
+                sum = assignProbabilities(edge.p->e.at(0), probs) + assignProbabilities(edge.p->e.at(1), probs);
+            }
+
+            probs.insert({edge.p, sum});
+
+            return ComplexNumbers::mag2(edge.w) * sum;
+        }
+
+    public:
+        char measureOneCollapsing(vEdge& root_edge, const Qubit index, const bool assume_probability_normalization, std::mt19937_64& mt, fp epsilon = 0.001) {
```

Comment:
```suggestion
        char measureOneCollapsing(vEdge& rootEdge, const Qubit index, const bool assumeProbabilityNormalization, std::mt19937_64& mt, fp epsilon = 0.001) {
```
Isn't probability normalisation always assumed?